src/transformers/models/whisper/generation_whisper.py

```diff
@@ -1167,49 +1190,47 @@ def replace_or_add(lst: List[int], num: int, itr: Iterator[int]):
                 num_segment_frames=num_segment_frames,
             )
 
-            if torch.unique(lang_ids).shape[0] > 1:
-                raise ValueError(
-                    "Multiple languages detected when trying to predict the most likely target language for transcription. It is currently not supported to transcribe to different languages in a single batch. Please make sure to either force a single language by passing `language='...'` or make sure all input audio is of the same language."
-                )
-
-            lang_id = lang_ids[0].item()
-
             # append or replace lang_id to init_tokens
-            if len(init_tokens) > 1:
-                init_tokens[1] = lang_id
-            else:
-                init_tokens.append(lang_id)
-
-        if task is not None:
-            if task in TASK_IDS:
-                init_tokens.append(generation_config.task_to_id[generation_config.task])
-                task_id = generation_config.task_to_id[generation_config.task]
+            for i in range(len(init_tokens)):
```

Comment:
`init_tokens` is only 1 long in the case of `language=None` at this point (because when you do the deepcopy on init_tokens, it assumes that a list of languages has been passsed in). We also need to set language for a later check in the function, since the dim of language is used in a check

To fix this, we can do:
```
id_to_lang = {v: k for k, v in generation_config.lang_to_id.items()}
language = [id_to_lang[lang_id.item()] for lang_id in lang_ids]
init_tokens = [copy.deepcopy(init_tokens[0]) for _ in lang_ids]
```

(the init tokens is a bit janky above, it would be cleaner if we didn't do the deepcopy above since then it is a list of lists).

Reply:
you can validate the current failure with the test of multiple languages, but passing `None` as the language

Reply:
True! But it's becoming a bit convoluted then. It's probably cleaner to revert 0b424f7 so that `init_tokens` and `language` are always of length `batch_size`. @ylacombe?

Reply:
Not sure to follow the issue here, could you expand on this ?

Reply:
@ylacombe 0b424f7 was based on [your suggestion](https://github.com/huggingface/transformers/pull/29688#discussion_r1528385312), but it turns out it broke language detection. So maybe we can revert that unless you think it would introduce significant overhead.

Another option would be to do something like this:
```python
        if isinstance(language, (list, tuple)):
            # ...
        elif language is None:
            # Language will be detected for each item in batch
            language = [None] * batch_size
        else:
            language = [language]  # Use a length-1 list now, broadcast later

        # Separate init_tokens for each language
        init_tokens = [copy.deepcopy(init_tokens) for _ in language]
```

Reply:
Actually @naveen-corpusant, I don't think you're right on your second point. That later check is not checking the detected language, it's there to check if language was passed to `generate()` and if so, set the task to transcription. I'll introduce a variable `languages` in order to keep the original `language` parameter intact.

Reply:
I went with my suggestion above, so `language` will be a length-1 list if a single language was passed, otherwise it will be a length-`batch_size` list of either languages or `None`s (if the languages need to be detected).